integration/spark/src/main/scala/org/apache/spark/sql/CarbonCountStar.scala

```diff
@@ -108,4 +125,145 @@ case class CarbonCountStar(
+      for (validSegmentPath <- validSegmentPaths) {
+        if (tableSegmentIndexes.get(validSegmentPath) == null) {
+          val segmentIndexMeta = new SegmentIndexMeta(validSegmentPath)
+          tableSegmentIndexes.put(validSegmentPath, segmentIndexMeta)
+        }
+      }
+
+      val numThreads = Math.min(Math.max(validSegmentPaths.length, 1), 4)
+      val executorService = Executors.newFixedThreadPool(numThreads)
+      // to get the index files of valid segments from cache or scan through each segment
+      val futures = new java.util.ArrayList[Future[Long]]
+      for (segmentPath <- validSegmentPaths) {
+        futures.add(executorService.submit(new Callable[Long] {
+          override def call(): Long = {
+            var rowCountCurrentSeg: Long = 0
+            var indexFilesCurrentSeg: java.util.List[CarbonFile] = null;
+            // tableSegmentIndexes already init.
+            if (tableSegmentIndexes.get(segmentPath) != null &&
+              tableSegmentIndexes.get(segmentPath).getSegmentIndexFiles.size() > 0) {
+              indexFilesCurrentSeg =
+                tableSegmentIndexes.get(segmentPath).getSegmentIndexFiles
+            } else {
+              // read from seg file
+              val sfs = new SegmentFileStore(carbonTable.getTablePath, segmentPath)
+              val indexFileStore = new SegmentIndexFileStore
+              indexFilesCurrentSeg = indexFileStore.getAllIIndexOfSegment(sfs.getSegmentFile,
+                carbonTable.getTablePath, SegmentStatus.SUCCESS, false)
+              // tableSegmentIndexes already init and segmentIndexMeta not null.
+              val segmentIndexMeta = tableSegmentIndexes.get(segmentPath)
+              segmentIndexMeta.addSegmentIndexFiles(indexFilesCurrentSeg)
+              // cache all the index files of this segment
+              tableSegmentIndexes.put(segmentPath, segmentIndexMeta)
+            }
+
+            // use partition prune to get row count of each pruned index file and
+            // cache it for this segment
+            var prunedIndexFilesCurrentSeg: java.util.List[String] =
+            new java.util.ArrayList[String]()
+            for (indexFile <- indexFilesCurrentSeg.asScala) {
+              // check whether the index files located in pruned partition
+              val formattedPath = indexFile.getAbsolutePath
+                .replace("\\", "/")
+              if (prunedPartitionPaths.contains(
+                formattedPath.substring(0, formattedPath.lastIndexOf("/")))) {
+                prunedIndexFilesCurrentSeg.add(formattedPath)
+              }
+            }
+            // get the row count from cache or read it from pruned index files
+            var toReadIndexFiles: java.util.List[String] = new java.util.ArrayList[String]()
+            prunedIndexFilesCurrentSeg.asScala.foreach(prunedIndexFilePath =>
+              if (tableSegmentIndexes.get(segmentPath).getPrunedIndexFileToRowCountMap
```

Comment:
refactor and assign to  variable for `tableSegmentIndexes.get(segmentPath).getPrunedIndexFileToRowCountMap .get(prunedIndexFilePath)` and to update in `SegmentIndexMeta` add a setter for map